solutionbox/code_free_ml/mltoolbox/code_free_ml/data/trainer/task.py

```diff
@@ -0,0 +1,849 @@
+          gfile.Copy(source, dest_absolute)
+
+    # only keep the last 3 models
+    saved_model_export_utils.garbage_collect_exports(
+        export_dir_base,
+        exports_to_keep=3)
+
+    # save the last model to the model folder.
+    # export_dir_base = A/B/intermediate_models/
+    if keep_target:
+      final_dir = os.path.join(args.job_dir, 'evaluation_model')
+    else:
+      final_dir = os.path.join(args.job_dir, 'model')
+    if file_io.is_directory(final_dir):
+      file_io.delete_recursively(final_dir)
+    file_io.recursive_create_dir(final_dir)
+    _recursive_copy(export_dir, final_dir)
+
+    return export_dir
+
+  if keep_target:
+    intermediate_dir = 'intermediate_evaluation_models'
+  else:
+    intermediate_dir = 'intermediate_prediction_models'
+
+  return export_strategy.ExportStrategy(intermediate_dir, export_fn)
+
+
+# TODO(brandondutra) add this to tft, and then remove it.
+def build_csv_transforming_training_input_fn(raw_metadata,
+                                             transform_savedmodel_dir,
+                                             raw_data_file_pattern,
+                                             training_batch_size,
+                                             raw_keys,
+                                             transformed_label_keys,
+                                             convert_scalars_to_vectors=True,
+                                             num_epochs=None,
+                                             randomize_input=False,
+                                             min_after_dequeue=1,
+                                             reader_num_threads=1):
+  """Creates training input_fn that reads raw csv data and applies transforms.
+
+  Args:
+    raw_metadata: a `DatasetMetadata` object describing the raw data.
+    transform_savedmodel_dir: a SavedModel directory produced by tf.Transform
+      embodying a transformation function to be applied to incoming raw data.
+    raw_data_file_pattern: List of files or pattern of file paths containing
+        `Example` records. See `tf.gfile.Glob` for pattern rules.
+    training_batch_size: An int or scalar `Tensor` specifying the batch size to
+      use.
+    raw_keys: List of string keys giving the order in the csv file.
+    transformed_label_keys
+    convert_scalars_to_vectors: Boolean specifying whether this input_fn should
+      convert scalars into 1-d vectors.  This is necessary if the inputs will be
+      used with `FeatureColumn`s as `FeatureColumn`s cannot accept scalar
+      inputs. Default: True.
+
+    num_epochs
+    randomize_input
+    min_after_dequeue
+    reader_num_threads
+    queue_capacity
+
+  Returns:
+    An input_fn suitable for training that reads raw csv training data and
+    applies transforms.
+
+  """
+
+  if not raw_keys:
+    raise ValueError("raw_keys must be set.")
+
+  column_schemas = raw_metadata.schema.column_schemas
+
+  # Check for errors.
+  for k in raw_keys:
```

Comment:
No. Yes. TFT style input readers in mappers.py allow for only using a subset of the data.